backend/common_test.go

```diff
@@ -51,14 +128,86 @@ func TestDataSourceInstanceSettings(t *testing.T) {
 						User:     "user",
 						Password: "pwd",
 					},
+					Labels: map[string]string{
+						"datasource_name": "ds2",
+						"datasource_uid":  "uid2",
+					},
+					CustomOptions: map[string]interface{}{
+						dataCustomOptionsKey: map[string]interface{}{},
+						secureDataCustomOptionsKey: map[string]string{
+							"basicAuthPassword": "bpwd",
+							"password":          "pwd",
+						},
+					},
+				},
+			},
+			{
+				instanceSettings: &DataSourceInstanceSettings{
+					JSONData: []byte("{ \"key\": \"value\" }"),
+					DecryptedSecureJSONData: map[string]string{
+						"sKey": "sValue",
+					},
+				},
+				expectedClientOptions: httpclient.Options{
+					CustomOptions: map[string]interface{}{
+						dataCustomOptionsKey: map[string]interface{}{
+							"key": "value",
+						},
+						secureDataCustomOptionsKey: map[string]string{
+							"sKey": "sValue",
+						},
+					},
 				},
 			},
 		}
 
 		for _, tc := range tcs {
 			opts, err := tc.instanceSettings.HTTPClientOptions()
 			assert.NoError(t, err)
-			assert.Equal(t, tc.expectedClientOptions.BasicAuth, opts.BasicAuth)
+			if tc.expectedClientOptions.BasicAuth != nil {
+				assert.Equal(t, tc.expectedClientOptions.BasicAuth, opts.BasicAuth)
+			} else {
+				assert.Nil(t, opts.BasicAuth)
+			}
+
+			if tc.expectedClientOptions.Labels != nil {
+				assert.Equal(t, tc.expectedClientOptions.Labels, opts.Labels)
+			}
+
+			jsonData := JSONDataFromHTTPClientOptions(opts)
+			expectedJSONData := JSONDataFromHTTPClientOptions(tc.expectedClientOptions)
+			secureJSONData := SecureJSONDataFromHTTPClientOptions(opts)
+			expectedSecureJSONData := SecureJSONDataFromHTTPClientOptions(tc.expectedClientOptions)
+
+			if len(tc.expectedClientOptions.CustomOptions) > 0 {
+				assert.Equal(t, tc.expectedClientOptions.CustomOptions, opts.CustomOptions)
+				assert.Equal(t, expectedJSONData, jsonData)
+				assert.Equal(t, expectedSecureJSONData, secureJSONData)
+			} else {
+				assert.Empty(t, opts.CustomOptions)
+				assert.Empty(t, jsonData)
+				assert.Empty(t, secureJSONData)
+			}
 		}
 	})
 }
+
+func TestCustomOptions(t *testing.T) {
+	opts := &httpclient.Options{}
+	expectedJSONData := map[string]interface{}{
+		"key": "value",
+	}
+	expectedSecureJSONData := map[string]string{
+		"sKey": "sValue",
+	}
+	setCustomOptionsFromHTTPSettings(opts, &HTTPSettings{
+		JSONData:       expectedJSONData,
+		SecureJSONData: expectedSecureJSONData,
+	})
+
+	jsonData := JSONDataFromHTTPClientOptions(*opts)
+	secureJSONData := SecureJSONDataFromHTTPClientOptions(*opts)
+
+	require.Equal(t, expectedJSONData, jsonData)
+	require.Equal(t, expectedSecureJSONData, secureJSONData)
+}
```

Comment:
I'd also add cases where the `JSONData` isn't actual json, or can be an invalid one and assert throwing an error.

Reply:
Good point. Fixed